---
title: "README_marathonas_validation"
author: "Robin Rohwer"
date: "1/22/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



## Reproduce TaxAss validation using full-length 16S assignments

This readme walks you through the data processing to reproduce "simulated 
tag data" from full-length 16S sequences. This was used to validate TaxAss.

The process:  

* Start with full length 16S sequences from Marathonas reservoir in Greece.   
    * These do not already exist in the FreshTrain database.  
* Align the full length sequences to the FreshTrain manually using Arb.   
    * The result of this is zipped in this folder. Likely chimeras were removed.  
* Trim the full-length Marathonas sequences to a primer region of choice.   
    * That's where this Readme picks up.  
* Assign taxonomy to the short primer regions using TaxAss.  
    * When I say "tax!", you say "ass!"  
* Compare the TaxAss assignment results to the "true" classifications known 
from the full length alignments.




## Unzip and add TaxAss tax-scripts  

Start with `Marathonas_test` as your present working directory.  

Add the tax-scripts and taxonomy database files to the `add-tax-scripts-and-databases` folder:  

```{bash, eval=F}
cp ../tax-scripts/*  add-tax-scripts-and-databases
cp ../FreshTrain-files/whatever 
unzip them
# add comprehensive database also
```  




## Trim full-length sequences to different V regions  

Trim the Marathonas full-length 16S sequences to the variable regions amplified by primers in tag data. Do this for several commonly used primer sets that amplify regions of differing lengths. Put
each V-region analysis in a different folder.

This uses the shell script `trim_full-length_to_primer_region.sh` and requires the program mothur to be on your computer's path. The syntax to call the script is:  
```
./trim_full-length_to_primer_region.sh primers.oligos aligned-fasta-without-filetype name-for-trimmed.fasta
```

### V4 region
```{bash, eval=F}  
mkdir v4_tags
# add all the scripts and reference files
cp add-tax-scripts-and-databases/* v4_mara/
cp oligo-files/v4_515F-806R.oligos v4_mara/
cp oligo-files/ecoli_16S.fasta v4_mara/
cp trim_full-length_to_primer_region.sh v4_mara/
# add all the marathonas data files
cp mara_aligned.fasta v4_mara/
cp mara_dummy.abund v4_mara/otus.abund
# run the script to trim the fasta file to primer region
cd v4_mara
./trim_full-length_to_primer_region.sh v4_515F-806R.oligos mara_aligned mara_v4.fasta
```




## TaxAss the simulated tag data  

Run TaxAss on each simulated tag dataset. 

### V4 region
```{bash, eval=F}
mkdir mara_v4
cp mara_v4.fasta mara_v4/otus.fasta
cp add-tax-scripts-and-databases/* mara_v4
cd mara_v4
./RunSteps_1-14.sh > terminal_output_1-14.txt
./RunStep_15.sh > terminal_output_15.txt
# don't ./RunStep_16.sh because need some intermediate files for analysis
```




## Compare TaxAss names to full-length names  

Using script `TaxAss_full-length_16S.R`  


and this doesn't have command line entry yet.  

