---
title: "How to format the ARB-exported FreshTrain for use with TaxAss"
author: "Robin Rohwer"
date: "last updated Feb 1, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview of steps

* export all full-length 16S from ARB  
* reformatting to make taxonomy file compatible with mothur  
* remove references lacking FreshTrain names  
* adjust coarse-level nomenclature (phylum/class/order) to match comprehensive database of choice

## ARB export  

The FreshTrain comes from an ARB database of 16S clone libraries from freshwater systems. 
All of the clones were aligned in ARB alongside the Greengenes database to determine the 
phylogeny of abundant freshwater taxa introduced in [Newton *et al.* 2011](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3063352/). 
Currently all full-length 16S sequences are exported from ARB, and this includes some of 
the freshwater clones that do not have "FreshTrain names," i.e. they were found in a freshwater
sample but they are not one of the abundant freshwater taxa included in the Newton nomenclature.  

## Reformatting  


## Remove non-FreshTrain refs

Before taxonomy assignment with the FreshTrain and TaxAss, these sequences need to be removed 
because only relevant reference sequences should exist for the percent identity calculation.

## Adjust coarse-level nomenclature  



